server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBFunctionService.java

```diff
@@ -41,33 +41,37 @@ public class DBFunctionService {
     private ConnectConsoleService consoleService;
 
     public List<DBFunction> list(ConnectionSession connectionSession, String dbName) {
-        DBSchemaAccessor accessor = DBSchemaAccessors.create(connectionSession);
-        return accessor.listFunctions(dbName).stream().map(item -> {
-            DBFunction function = new DBFunction();
-            function.setFunName(item.getName());
-            function.setErrorMessage(item.getErrorMessage());
-            function.setStatus(item.getStatus());
-            return function;
-        }).collect(Collectors.toList());
+        return connectionSession.getSyncJdbcExecutor(
+                ConnectionSessionConstants.BACKEND_DS_KEY)
+                .execute((ConnectionCallback<List<DBPLObjectIdentity>>) con -> getFunctionExtensionPoint(
+                        connectionSession).list(con, dbName))
+                .stream().map(
+                        item -> {
+                            DBFunction function = new DBFunction();
+                            function.setFunName(item.getName());
+                            function.setErrorMessage(item.getErrorMessage());
+                            function.setStatus(item.getStatus());
+                            return function;
+                        })
+                .collect(Collectors.toList());
     }
 
     public DBFunction detail(ConnectionSession connectionSession, String dbName, String funName) {
-        DBSchemaAccessor accessor = DBSchemaAccessors.create(connectionSession);
-        return accessor.getFunction(dbName, funName);
+        return connectionSession.getSyncJdbcExecutor(
+                ConnectionSessionConstants.BACKEND_DS_KEY)
+                .execute((ConnectionCallback<DBFunction>) con -> getFunctionExtensionPoint(connectionSession)
+                        .getDetail(con, dbName, funName));
     }
 
     public ResourceSql getCreateSql(@NonNull ConnectionSession session,
             @NonNull DBFunction function) {
-        DBObjectTemplate<DBFunction> template;
-        if (session.getDialectType().isMysql()) {
-            template = new MySQLFunctionTemplate();
-        } else if (session.getDialectType().isOracle()) {
-            template = new OracleFunctionTemplate();
-        } else {
-            throw new UnsupportedOperationException("Unsupported dialect, " + session.getDialectType());
-        }
-        String ddl = template.generateCreateObjectTemplate(function);
-        return ResourceSql.ofSql(ddl);
+        return ResourceSql.ofSql(session.getSyncJdbcExecutor(
+                ConnectionSessionConstants.BACKEND_DS_KEY)
+                .execute((ConnectionCallback<String>) con -> getFunctionExtensionPoint(session)
+                        .generateCreateTemplate(function)));
     }
 
+    private FunctionExtensionPoint getFunctionExtensionPoint(@NonNull ConnectionSession session) {
+        return SchemaPluginUtil.getFunctionExtension(session.getDialectType());
```

Comment:
how to handle npe? `OdcPluginManager` get a logical flaw:
```Java
// OdcPluginManager#getSingleton
private <T extends ExtensionPoint> T getSingleton(List<T> collection, V object) {
        if (collection.size() > 1) {
            String message = MessageFormat.format("Expect single extension for {}, but got "
                    + "{}, extension : {}", object, collection.size(),
                    collection.stream().map(t -> t.getClass().getSimpleName()).collect(Collectors.toList()));
            throw new IllegalStateException(message);
        }
       // if collection's size is equal to 0?
        return collection.get(0);
    }
```
maybe we can solve this npe by enhancing the robustness of the `OdcPluginManager#getSingleton`

Reply:
👌 add code：
```if (collection.size() > 1) {
            String message = MessageFormat.format("Expect single extension for {}, but got "
                    + "{}, extension : {}", object, collection.size(),
                    collection.stream().map(t -> t.getClass().getSimpleName()).collect(Collectors.toList()));
            throw new IllegalStateException(message);
        } else if (collection.isEmpty()) {
            throw new UnsupportedException(String.format("Feature extension point is not supported for %s", object));
        }